### PR TITLE
[BUGFIX] Fix incorrect usage of "all()" in the comparison of validation results when executing an Expectation

### PR DESCRIPTION
**Pull request by alexsherstinsky**
…s when executing an Expectation.

Please annotate your PR title to describe what the PR does, then give a brief bulleted description of your PR below. PR titles should begin with [BUGFIX], [FEATURE],  [DOCS], or [MAINTENANCE]. If a new feature introduces breaking changes for the Great Expectations API or configuration files, please also add [BREAKING]. You can read about the tags in our [contributor checklist](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html).

Changes proposed in this pull request:
- JIRA: GE-320/GE-446
-
-

After submitting your PR, CI checks will run and @tiny-tim-bot will check for your CLA signature.

For a PR with nontrivial changes, we review with both design-centric and code-centric lenses.

In a design review, we aim to ensure that the PR is consistent with our relationship to the open source community, with our software architecture and abstractions, and with our users' needs and expectations. That review often starts well before a PR, for example in github issues or slack, so please link to relevant conversations in notes below to help reviewers understand and approve your PR more quickly (e.g. `closes #123`).

Previous Design Review notes:
-
-
-

### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/en/latest/contributing/style_guide.html?highlight=style%20guide)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/en/latest/contributing/contribution_checklist.html?highlight=checklist) of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added [unit tests](https://docs.greatexpectations.io/en/latest/contributing/testing.html#contributing-testing-writing-unit-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

Thank you for submitting!

### CODE REVIEW

File: great_expectations/core/expectation_validation_result.py [hunk 1]
```diff
@@ -96,7 +96,14 @@ def __eq__(self, other):
...
                     (self.result is None and other.result is None)
-                    or (all(self.result) == all(other.result)),
+                    or all(
+                        [
+                            self.result[k] == other.result[k]
+                            for k in set(self.result.keys()).intersection(
+                                other.result.keys()
+                            )
+                        ]
+                    ),
```
Comment (Shinnnyshinshin) [inline great_expectations/core/expectation_validation_result.py]: <3

**Comment by alexsherstinsky** [comment]
More investigation is needed.

File: great_expectations/core/expectation_validation_result.py [hunk 2]
```diff
@@ -79,6 +79,15 @@ def __eq__(self, other):
             # Delegate comparison to the other instance's __eq__.
             return NotImplemented
         try:
+            if self.result and other.result:
+                common_keys = set(self.result.keys()).intersection(other.result.keys())
```
Comment (cdkini) [inline great_expectations/core/expectation_validation_result.py]: Could this be simplified to 
```suggestion
                common_keys = self.result.keys() & other.result.keys()
```

Comment (alexsherstinsky) [reply great_expectations/core/expectation_validation_result.py]: @cdkini -- thanks! (will go in to the next commit).

File: great_expectations/core/expectation_validation_result.py (hunk 1 above)
Comment (alexsherstinsky) [reply great_expectations/core/expectation_validation_result.py]: @Shinnnyshinshin I updated this section further -- to make it more robust and easier to read.

File: tests/data_asset/test_expectation_decorators.py [hunk 3]
```diff
@@ -518,7 +519,7 @@ def expect_column_pair_values_to_be_different(
             "missing_count": 2,
             "unexpected_count": 3,
             "missing_percent": 40.0,
-            "unexpected_percent": 60.0,
+            "unexpected_percent": 100.0,
             "unexpected_percent_nonmissing": 100.0,
```
Comment (Shinnnyshinshin) [inline tests/data_asset/test_expectation_decorators.py:523]: Going through some changes, I noticed that some values have changed more than others... are these all "bugs" that we weren't able to catch before? Is there a pattern to what changed?

File: great_expectations/core/expectation_validation_result.py [hunk 4]
```diff
@@ -95,8 +105,7 @@ def __eq__(self, other):
                     # Result is a dictionary allowed to have nested dictionaries that are still of complex types (e.g.
                     # numpy) consequently, series' comparison can persist. Wrapping in all() ensures comparison is
                     # handled appropriately.
-                    (self.result is None and other.result is None)
-                    or (all(self.result) == all(other.result)),
+                    not (self.result or other.result) or contents_equal,
```
Comment (Shinnnyshinshin) [inline great_expectations/core/expectation_validation_result.py:108]: Thank you again for finding and correcting this

File: tests/data_asset/test_expectation_decorators.py (hunk 3 above)
Comment (alexsherstinsky) [reply tests/data_asset/test_expectation_decorators.py:523]: @Shinnnyshinshin Yes.  The previous values fail tests once the comparison has been fixed.

File: tests/expectations/core/test_expect_column_value_z_scores_to_be_less_than.py [hunk 5]
```diff
@@ -92,4 +144,30 @@ def test_spark_expect_column_value_z_scores_to_be_less_than_impl(
...
+            "missing_percent": 0.0,
+            "unexpected_percent_total": 0.0,
+            "unexpected_percent_nonmissing": 0.0,
+        },
+        exception_info={
+            "raised_exception": False,
+            "exception_traceback": None,
+            "exception_message": None,
+        },
+        meta={},
```
Comment (Shinnnyshinshin) [inline tests/expectations/core/test_expect_column_value_z_scores_to_be_less_than.py:172]: Just to highlight, this is a significant part that was missed by the bug in the `all() == all()` comparison

Comment (Shinnnyshinshin) [reply tests/expectations/core/test_expect_column_value_z_scores_to_be_less_than.py:172]: There exists a `pandas`, `sqlalchemy` and , `spark` version of this test.

**Review by Shinnnyshinshin — APPROVED** [review]
Thank you very very much for this. LGTM

**Pull request merged**